pep-0724.rst

```diff
@@ -0,0 +1,303 @@
+
+Given a ``TypeGuard`` and a calling function like so:
+
+.. code-block:: python
+
+    def guard(x: P) -> TypeGuard[R]: ...
+
+    def func1(val: A):
+        if guard(val):
+            reveal_type(val) # NP
+        else:
+            reveal_type(val) # NN
+
+Each use of a ``TypeGuard`` involves five types:
+
+* P = ``TypeGuard`` parameter type
+* R = ``TypeGuard`` return type
+* A = Argument type (pre-narrowed)
+* NP = Narrowed type (positive)
+* NN = Narrowed type (negative)
+
+
+This PEP proposes some modifications to :pep:`647` in order to allow a type
+checker to further narrow what a ``TypeGuard`` returns when the ``R`` type is
+consistent [#isconsistent]_ with the ``P`` type.
+
+* When the output type ``R`` of a ``TypeGuard`` is consistent [#isconsistent]_
+  with the type of its first input parameter (``P``), type checkers will apply
+  stricter type semantics:
+
+.. code-block:: python
+    
+    # Stricter TypeGuard possible (Kangaroo | Koala is consistent with Animal)
+    def is_marsupial(val: Animal) -> TypeGuard[Kangaroo | Koala]: 
+        return isinstance(val, Kangaroo | Koala)
+
+    # Stricter TypeGuard not possible (list[T] is not consistent with 
+    # list[T | None])
+    def has_no_nones(val: list[T | None]) -> TypeGuard[list[T]]: 
+        return None not in val
+
+* When stricter semantics are possible, the output type for the ``False`` or
+  ``NN`` case can be determined:
+
+.. code-block:: python
+
+    def is_str(val: str | int) -> TypeGuard[str]: # Stricter mode possible
+        return isinstance(val, str)
+
+    def func(val: str | int):
+        if is_str(val):
+            # Type checkers can assume val is a 'str' in this branch
+        else:
+            # Type checkers can assume val is an 'int' in this branch
+
+* When the output type of the ``TypeGuard`` is a union, the type checker can
+  apply additional type narrowing based on the type of the first input
+  argument, eliminating union elements that are impossible given the input
+  argument type:
+
+.. code-block:: python
+
+    def is_cardinal_direction(val: str) -> TypeGuard[Literal["N", "S", "E", "W"]]:
+        return val in ("N", "S", "E", "W")
+
+    def func(direction: Literal["NW", "E"]):
+        if is_cardinal_direction(direction):
+            # Literal["E"] The type cannot be "N", "S" or "W" here because of
+            # argument type
+        else:
+            # Literal["NW"]
+
+
+This logic can be summed up in a table:
+
+============ ======================= ===================
+|space|      Non strict TypeGuard    Strict TypeGuard
+============ ======================= ===================
+Applies when R not consistent with P R consistent with P
+NP is ..     R                       A & R
+NN is ..     A                       A & ~R
+============ ======================= ===================
+
+The definition of ``A & R`` and ``A & ~R`` would be left up to 
+the type checker itself as these concepts don't yet exist in Python typing.
```

Comment:
We're currently working on a definition of intersections prior to a PEP, and the question of Not also has come up over there. I would strongly suggest not leaving Not fully up to type checkers here\*, and to write this in a way where This should defer to a future intersection pep if this is ready before the intersection pep is. The intersection pep's current track is consistent with the behavior laid out as desirable here.

\* [This comment goes into some detail about what Not can be shown to be safely composed with.](https://github.com/python/typing/issues/801#issuecomment-1679883519 ) Notably, it hasn't been shown that type negation is safe to apply to gradual or unknown types. This PEP does not currently suggest to do this or allow this, but it should probably not introduce an internal type negation without forbiding type-checkers from expanding on this to prevent accidentally introducing something with unclear meaning in the process.

Reply:
Thanks that sounds like a much better idea than just leaving this vague. I'll reference the intersection design doc.

Reply:
I don't think a draft PEP should reference an incomplete design document. If and when the Intersection design reaches the point where it is submitted as a PEP, then it would be fine to provide a footnote in this PEP that references it. In the meantime, I think it's fine to say something to the effect that "future extensions to the type system may affect the implementation of this feature".

Reply:
I changed the wording but still left at least a potential for what might the change might be?

Reply:
I agree with Eric that it probably shouldn't reference the incomplete design, I only provided the link to be able to have it be confirmed that the work on intersections matches how you're using it here and for anyone who might be interested in exploring how the two in progress PEPs might interact in case people had concerns or additional feedback to either draft as a result.